mlibrary/lib.umich.edu
# Import Universal Header via CDN vs NPM

# Overview
The Universal Header has always had a visible/noticeable load on lib that we don’t see on any of the other sites where it’s added. This fix helps address the page jump on load.

Additionally, the UH is imported via NPM versus directly from the CDN. To align with how the design system is used elsewhere, this PR also changes the import to use the CDN. 

The noticeable page jump is due to the way Gatsby handles static server-side rendering of pages and client hydration. When a client visits a URL, they're given the server-built, static file--no JavaScript or interactivity outside of native HTML interaction. After the static file finishes loading, the page is hydrated with JavaScript. Since the Universal Header requires JavaScript to render, the `<m-universal-header>` component remains empty on the initial static file and only renders during hydration, hence the page jump.

Since the UH is the top-most element on the page, the jump is very noticeable. To fix the jump, I added an empty placeholder component at the top of the page. The placeholder has a `min-height` that is the same height as the UH and is the same `background-color` as the UH. I don't want to replicate the UH too much because any changes to the UH will need to be updated here, which seems counterintuitive to the centralized approach of the design system.

The UH then loads into the placeholder so the page doesn't jump. I'm also preloading the script to help get it loading faster. I don't know if that's doing much, though.

Finally, I created a `gatsby-shared.js` file and moved the `wrapPageElement` there. I then export it to the `gatsby-browser` and `gatsby-ssr` since I want these components to render on both the server and client.

> This pull request resolves [WEBSITE-154](https://mlit.atlassian.net/jira/software/projects/WEBSITE/boards/27?selectedIssue=WEBSITE-154).

## Anything else?
**Possible future enhancement**: Animate the height of the placeholder so any changes in height to the UH will ease the transition between the placeholder and UH. This could be a good solution for the alerts, since we're not currently doing anything for that. The alerts are nested within the UH so targeting them on Lib may be tricky. Perhaps this is an enhancement we can make on the design system's end?

## Testing
- Make sure the PR is consistent in these browsers:
- Run accessibility tests:
